quine-core/src/main/scala/com/thatdot/quine/graph/NodeActor.scala

```diff
@@ -139,6 +134,73 @@ private[graph] class NodeActor(
     case MergedHere(other) => !mergedIntoHere.contains(other)
   }
 
+  /**
+    * Drop duplicate events, and wrap the Events being kept in an `EventTime`.
+    *
+    * An event is considered a duplicate (and is dropped) if:
+    *  - the same event appears later in the sequence of events, or
+    *  - the event would have no effect when applied to this node.
+    *
+    * This implementation prioritizes avoiding heap allocations.
+    * Since it does a linear search of the later events,
+    * this could be less efficient for very large sequences of events.
```

Comment:
Found a bit more time: here's some more detail on the allocations:

The default implementation of `+:` (in 2.12, from `TraversibleLike`) will cause a fresh allocation for `laterEvents` each time. In practice, events tends to be a `List`, which has a more efficient `+:` `unapply`, but since the signature of the function is `Seq`, if the goal is to minimize allocations, I think this `+:` undermines the goal rather than supporting it (in general).

However, since `List` does have a 0-allocation destructor of the right shape (`::`, though `+:` works nearly as well), you could avoid the issue by redefining `accumulate` in terms of lists, passing in the initial argument with `toList`. This would also let you use a slightly "safer" `case Nil` instead of `case _` for the base case (the latter being more prone to future errors in refactoring)

Reply:
Rather than traversing the list from oldest to newest, and scanning the remainder of the list for any newer items of the same type on each entry, what about traversing the list once from newest to oldest, keeping (the first, newest) unique values, but ignoring repeats of the same type in the older entries?

Reply:
I think there is something in that. You don't really need to reverse the list at all, but the algorithm as it stands requires you to traverse from newest to oldest. Just copy it into some sort of mutable indexed Seq (an array would do) and just iterate through it in reverse.

You could even remove duplicate elements in place. It is awkward and expensive to do that in an array, but it might if duplicates are typically rare, then this would not be an issue, other than making the code a bit ugly.